snuba/query/processors/project_rate_limiter.py

```diff
@@ -0,0 +1,104 @@
+from snuba.query.matchers import String as StringMatch
+from snuba.query.matchers import Or, Param
+from snuba.request.request_settings import RequestSettings
+from snuba.state import get_configs
+from snuba.state.rate_limit import PROJECT_RATE_LIMIT_NAME, RateLimitParameters
+
+
+class ProjectRateLimiterProcessor(QueryProcessor):
+    """
+    If there isn't already a rate limiter on a project, search the top level conditions
+    for project IDs using the given project column name and add a rate limiter for them.
+    """
+
+    def __init__(self, project_column: str) -> None:
+        self.project_column = project_column
+
+    def process_query(self, query: Query, request_settings: RequestSettings) -> None:
+        # If the settings don't already have a project rate limit, add one
+        existing = request_settings.get_rate_limit_params()
+        for ex in existing:
+            if ex.rate_limit_name == PROJECT_RATE_LIMIT_NAME:
+                return
+
+        condition = query.get_condition_from_ast()
+        if not condition:
+            return
+
+        top_level = get_first_level_and_conditions(condition)
+        column_match = ColumnMatch(None, StringMatch(self.project_column))
+        match = Or(
+            [
+                FunctionCallMatch(
+                    StringMatch(ConditionFunctions.EQ),
+                    (column_match, Param("single", LiteralMatch(AnyMatch(int)))),
+                ),
+                FunctionCallMatch(
+                    StringMatch(ConditionFunctions.IN),
+                    (
+                        column_match,
+                        Param(
+                            "multiple",
+                            FunctionCallMatch(
+                                Or([StringMatch("array"), StringMatch("tuple")]),
+                                all_parameters=LiteralMatch(AnyMatch(int)),
+                            ),
+                        ),
+                    ),
+                ),
+            ]
+        )
+
+        project_id = 0  # TODO rate limit on every project in the list?
+        for cond in top_level:
+            result = match.match(cond)
+            if result is not None:
+                if result.contains("single"):
+                    exp = result.expression("single")
+                    if isinstance(exp, Literal) and isinstance(exp.value, int):
+                        project_id = exp.value
+                        break
+                elif result.contains("multiple"):
+                    exp = result.expression("multiple")
+                    assert isinstance(exp, FunctionCall)
+                    found = False
+                    for p in exp.parameters:
+                        if isinstance(p, Literal) and isinstance(p.value, int):
+                            project_id = p.value
+                            found = True
+                            break
+
+                    if found:
+                        break
+
```

Comment:
There is already an api to do this.
https://github.com/getsentry/snuba/blob/master/snuba/clickhouse/query_dsl/accessors.py#L29-L31

That one is not specifically looking for top level conditions as the goal is to find all project ids mentioned in the conditions, which seems a valid behavior for this use case as well.

Reply:
That function also takes into account OR conditions, which I don't think we want. At least, the existing project rate limiter does not do that.

Reply:
Taking OR into account is not a problem. 
The project_id is a mandatory column in the entities that would have this processor. This means that whatever project you may have stated in an OR condition, to be meaningful, must also be in a top level AND condition. 
If your query requires a condition like
`project_id IN (1,2)` to pass validation
Then you may have a 
`project_id = 5` in an OR, but your query would be already conceptually broken, even if we apply the project 5 rate lim it as well, it is not an issue.

Also this logic is already broken as it is because it picks an arbitrary project from the extension to apply the rate limit when you could provide multiple projects in the extension.

Reply:
I just wanted to be clear that if I change this it will not behave the same way that the project extension does. Otherwise, I am all about removing duplicate code.